### ISSUE TITLE
Can't compile this tool/code...help please

### ISSUE THREAD

**Issue by serine** [deeptools/Knight-Ruiz-Matrix-balancing-algorithm#20]
Hi there, 

I was wondering if this is a working implementation of KR?  I have a few samples in spares (COO) format that I'd like to KR normalised. I prefer not to go via other tools e.g HiCExplore as I couldn't see an easy way to read sparse matrix in. 

I've trying building the source, but I think there is some issue with the code..

Here 

```
krbalancing.cpp: In member function ‘void kr_balancing::computeKR()’:                                                                                                                                              
krbalancing.cpp:56:47: error: no match for ‘operator-’ (operand types are ‘const ConstantReturnType {aka const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, 1> >}’ 
and ‘SparseMatrixCol {aka Eigen::SparseMatrix<double, 0, long int>}’)                                                                                                                                              
   rk = Eigen::VectorXd::Constant(v.rows(), 1) - v; //Vecotr of 1 - v 
```

I tried building with `python setup.py install` inside virtualenv with ` pip install pybind11`. and directly with 

```
c++ -O3 -Wall -I /path/to/eigen -shared -std=c++11 -fPIC $(python3 -m pybind11 --includes) KRBalancing.cpp -o KRBalancing$(python3-config --extension-suffix)
```

I also had to add this line into `setup.py` cause header files weren't getting find.

```
  include_dirs=[
                          # Path to eigen3 headers
                          get_include(),
                          "/home/kirillt/downloads/Knight-Ruiz-Matrix-balancing-algorithm/venv/include/site/python3.6"
                      ],
``` 
I think `site/python3.6` is confusing things??

Let me know if I'm doing something wrong and/or if I can even use this tool. 

Thanks ! 

p.s attached full log from `python setup.py install`

[log.txt](https://github.com/deeptools/Knight-Ruiz-Matrix-balancing-algorithm/files/5412401/log.txt)

**Comment by serine** [comment]
also I had to install `libeigen3-dev`, which there was no mention anywhere as a dependency, which needs `sudo`, not sure if there is a work around via conda

```
apt install libeigen3-dev
```

**Comment by joachimwolff** [comment]
You usually don't need anything from your systems package manager, it works with non-root privileges and conda dependencies. However, a 2 minutes try to compile it on my system failed, so it seems we have not fully given the correct dependency list. @LeilyR Can you have a look at this? 

In general: you can install via conda and the bioconda channel the package: `conda install krbalancing -c bioconda -c conda-forge` and import it to python For example a code snipped from https://github.com/deeptools/HiCExplorer/blob/master/hicexplorer/hicCorrectMatrix.py#L729

Please note that ma.matrix is a csr_matrix.

```
from krbalancing import *
kr = kr_balancing(ma.matrix.shape[0], ma.matrix.shape[1],
                              ma.matrix.count_nonzero(),
                              ma.matrix.indptr.astype(np.int64, copy=False),
                              ma.matrix.indices.astype(np.int64, copy=False), 
                              ma.matrix.data.astype(np.float64, copy=False))
log.debug('passed pointers')
kr.computeKR()
# set it to False since the vector is already normalised
# with the previous True
# correction_factors = np.true_divide(1, kr.get_normalisation_vector(False).todense())
correction_factors = kr.get_normalisation_vector(False).todense()

if args.outFileName.endswith('.h5'):
    corrected_matrix = kr.get_normalised_matrix(True)
```

Best, 

Joachim

**Comment by serine** [comment]
@joachimwolff This is great, thanks for quick response. I didn't know you had packaged it in conda, that simplifies things a lot for me. 

Regarding actual code, I'm struggling to get it to work the way I expect.. (perhaps I'm missing something)

Here is my input table (10 lines) 

```
3100000 3150000 407
3100000 3200000 170
3100000 3250000 69
3100000 3300000 45
3100000 3350000 36
3100000 3400000 24
3100000 3450000 64
3100000 3500000 36
3100000 3550000 33
3100000 3600000 26
```

This is what goes in and comes out

```
MSG: Data in -> [407 170  69  45  36  24  64  36  33  26]
MSG: Data in len -> 10
MSG: Computing ..
normalisation factor is 41.1052
MSG: Norm data out -> [0.00059184 0.00059184 0.00059184 ... 0.         0.         0.        ]
MSG: Norm data out len -> 3100011
```

I understand that `numpy/scipy` for some reason starts everything at zero? I basically expect 10 elements to go in and 10 to come out... 

here is my code

```python
#!/usr/bin/env python3

import sys
import numpy as np
import pandas as pd
from scipy.sparse import csr_matrix
import krbalancing

fn_in = sys.argv[1]
fn_out = fn_in.replace(".txt", "_normed.txt")

dat1 = pd.read_table(fn_in, header=None, names=["region1", "region2", "IF"])

mat1 = csr_matrix((dat1.iloc[:,2].to_numpy(), (dat1.iloc[:,0].to_numpy(), dat1.iloc[:,1].to_numpy())))

print("MSG: Data in -> %s" % mat1.data)
print("MSG: Data in len -> %s" % len(mat1.data))

kr = krbalancing.kr_balancing(mat1.shape[0], mat1.shape[1],
                              mat1.count_nonzero(),
                              mat1.indptr.astype(np.int64, copy=False),
                              mat1.indices.astype(np.int64, copy=False),
                              mat1.data.astype(np.float64, copy=False))

print("MSG: Computing ..", file = sys.stderr)

kr.computeKR()
#correction_factors = kr.get_normalisation_vector(False).todense()

mat1_norm = kr.get_normalised_matrix(True)

print("MSG: Norm data out -> %s" % mat1_norm.data)
print("MSG: Norm data out len -> %s" % len(mat1_norm.data))

````

Thanks a lot in advance !  


 p.s i basically gonna write normalised table as three column tsv at the end

**Comment by LeilyR** [comment]
interesting, I can already see what is happening here. The indices in your example started from `3100000` and it added 10 into that, so you got `3100011` for length. The actual length is indeed 10. This should be scipy behavior, I need to look into my code though, to say something in certain. However this should not affect your normalization I would say.  You still got the matrix you looked for it, right? (mat1_norm.data), maybe you can consider changing your input indices and their pointers.

**Comment by serine** [comment]
@LeilyR  thanks for your response. No I haven't got the desired results yet unfortunately. I'm hoping to end up with another three column tsv file with normalised values. I'm aiming to use [Armatus TAD caller](https://github.com/kingsfordgroup/armatus) on that file. 

when I attempt to write the matrix out with this call

```python
np.savetxt("check.tsv", mat1_norm, delimiter = ",")
```

I get this traceback

```python
Traceback (most recent call last):
  File "./kr_norm.py", line 35, in <module>
    np.savetxt("check.mat", mat1_norm, delimiter = ",")
  File "<__array_function__ internals>", line 5, in savetxt
  File "/home/kirillt/.miniconda/envs/kr_norm/lib/python3.8/site-packages/numpy/lib/npyio.py", line 1371, in savetxt
    raise ValueError(
ValueError: Expected 1D or 2D array, got 0D array instead
```

I'm not super familiar with numpy/scipy libraries, but this is why I was asking about the matrix dimensions - they change 

```python
MSG: Data in shape -> <bound method spmatrix.get_shape of <3100001x3600001 sparse matrix of type '<class 'numpy.int64'>'
        with 10 stored elements in Compressed Sparse Row format>>


MSG: Norm data shape -> <bound method spmatrix.get_shape of <3100001x3600001 sparse matrix of type '<class 'numpy.float64'>'
        with 3100011 stored elements in Compressed Sparse Column format>>
```

 I don't know if I need to make call to drop zero values or something..but like I said I was hoping to get the same matrix shape out. 

Thanks !

**Comment by serine** [comment]
I also tried turning `mat1_norm` into numpy array with `.toarray()` call, like this 

```python
np.savetxt("check.tsv", mat1_norm.toarray(), delimiter = ",")
```

and I get this traceback

```python
Traceback (most recent call last):
  File "./kr_norm.py", line 37, in <module>
    np.savetxt("check.mat", mat1_norm.toarray(), delimiter = ",")
  File "/home/kirillt/.miniconda/envs/kr_norm/lib/python3.8/site-packages/scipy/sparse/compressed.py", line 1025, in toarray
    out = self._process_toarray_args(order, out)
  File "/home/kirillt/.miniconda/envs/kr_norm/lib/python3.8/site-packages/scipy/sparse/base.py", line 1185, in _process_toarray_args
    return np.zeros(self.shape, dtype=self.dtype, order=order)
MemoryError: Unable to allocate 81.2 TiB for an array with shape (3100001, 3600001) and data type float64
```

but I think the actual issue is unwanted number of elements (that I don't get how to drop..)

**Comment by LeilyR** [comment]
This is what you can do for the 10 lines text file you have sent, although bear in mind that for your real matrix you need to reshape. 
```

dat1 = pd.read_csv("matrix.tsv", sep = "\t", header = None)
print(dat1)
mat1 = csr_matrix(dat1[2].values)
print(mat1)

print("MSG: Data in -> %s" % mat1.data)
print("MSG: Data in len -> %s" % len(mat1.data))

kr = krbalancing.kr_balancing(mat1.shape[0], mat1.shape[1],
                              mat1.count_nonzero(),
                              mat1.indptr.astype(np.int64, copy=False),
                              mat1.indices.astype(np.int64, copy=False),
                              mat1.data.astype(np.float64, copy=False))

kr.computeKR()
mat1_norm = kr.get_normalised_matrix(True)
print("MSG: Norm data out -> %s" % mat1_norm.data)
print("MSG: Norm data out len -> %s" % len(mat1_norm.data))
output_data = dat1
output_data[2] = mat1_norm.data
output_data.to_csv("output.tsv", sep = "\t", header = None, index = False)
```
output would be:
```
         0        1    2
0  3100000  3150000  407
1  3100000  3200000  170
2  3100000  3250000   69
3  3100000  3300000   45
4  3100000  3350000   36
5  3100000  3400000   24
6  3100000  3450000   64
7  3100000  3500000   36
8  3100000  3550000   33
9  3100000  3600000   26
  (0, 0)	407
  (0, 1)	170
  (0, 2)	69
  (0, 3)	45
  (0, 4)	36
  (0, 5)	24
  (0, 6)	64
  (0, 7)	36
  (0, 8)	33
  (0, 9)	26
MSG: Data in -> [407 170  69  45  36  24  64  36  33  26]
MSG: Data in len -> 10
MSG: Norm data out -> [1413.00002326    0.            0.            0.            0.
    0.            0.            0.            0.            0.        ]
MSG: Norm data out len -> 10
```
output_data: 
```
3100000	3150000	1413.0000232648774
3100000	3200000	0.0
3100000	3250000	0.0
3100000	3300000	0.0
3100000	3350000	0.0
3100000	3400000	0.0
3100000	3450000	0.0
3100000	3500000	0.0
3100000	3550000	0.0
3100000	3600000	0.0
```

**Comment by serine** [comment]
@LeilyR thanks a lot for helping, but I'm not 100% sure it is working the right way and also I'm little confused by what you mean that "your real matrix you need to reshape"

Running your code on the sample data (10 lines) gives the same results as you've pasted above, this is great. But when I ran the same code on the actual data set, I only got a single non zero value for on the coordinates pair, I don't this this is right..

```BASH
+(kr_norm) [bioplat1]~/.../results (master)$ grep -v -P "0.0$" 1_normed.txt 
3000000 3050000 29957877.651739568

+(kr_norm) [bioplat1]~/.../results (master)$ wc -l 1_normed.txt 
4018619 1_normed.txt
```

I think this is due to the fact that you construct `csr_matrix` like this `mat1 = csr_matrix(dat1[2].values)` which has `mat1.shape` `(1, 10)`

The true shape of this matrix is actually 11 x 11

```python
#!/usr/bin/env python3

import sys
import pandas as pd

fn_in = sys.argv[1]

print("MSG: Reading -> %s" % fn_in, file = sys.stderr)

dat1 = pd.read_table(fn_in, header=None, names=["region1", "region2", "IF"])

res1 = dat1.iloc[:,0].to_list()
res1 += dat1.iloc[:,1].to_list()

res2 = sorted(set(res1))

print(res2)
print(len(res2))
```

that gives this results 

```python
[3100000, 3150000, 3200000, 3250000, 3300000, 3350000, 3400000, 3450000, 3500000, 3550000, 3600000]
11
```
I don't know any C++ but from looking at the code for a while I think this line of code [`A.resize(input_rows, input_cols);`](https://github.com/deeptools/Knight-Ruiz-Matrix-balancing-algorithm/blob/572c865836d54c64bf5def9a988c493488f6b062/src/krbalancing.cpp#L10) should be `A.resize(11, 11)` instead of `A.resize(1, 10)` ? Those are passed in via `mat1.shape[0], mat1.shape[1],` I think this is the call that converts sparse matrix into full, square matrix... I could be wrong here, but either way I don't seem to be the right answer..

Thanks again

**Comment by serine** [comment]
I expect the full matrix to look like this 

```R
> dat <- read.table("~/Downloads/tt.txt")
> HiCcompare::sparse2full(dat)

Matrix dimensions: 11x11
        3100000 3150000 3200000 3250000 3300000 3350000 3400000 3450000 3500000 3550000 3600000
3100000       0     407     170      69      45      36      24      64      36      33      26
3150000     407       0       0       0       0       0       0       0       0       0       0
3200000     170       0       0       0       0       0       0       0       0       0       0
3250000      69       0       0       0       0       0       0       0       0       0       0
3300000      45       0       0       0       0       0       0       0       0       0       0
3350000      36       0       0       0       0       0       0       0       0       0       0
3400000      24       0       0       0       0       0       0       0       0       0       0
3450000      64       0       0       0       0       0       0       0       0       0       0
3500000      36       0       0       0       0       0       0       0       0       0       0
3550000      33       0       0       0       0       0       0       0       0       0       0
3600000      26       0       0       0       0       0       0       0       0       0       0
```

**Comment by serine** [comment]
just in case you'll be wondering `HiCcompare::KRnorm()` has an [R implementation here](https://github.com/dozmorovlab/HiCcompare/blob/23560676a057f1051f5ba5666baf6f6a2362b0ae/R/KRnormalization.R#L16) but it's very slow and/or it gets stuck in one of the `while` loops (don't know why), hence why I'm looking for fast and alternative implementation of `KR` normalisation (I have fair few samples). Thanks

**Comment by LeilyR** [comment]
that is exactly what i meant by reshaping. You want to call a 11 by 11 matrix from a one column text file , this is not going to work unless you take care of it before correction.
Example:
```
dat1 = pd.read_csv("matrix.tsv", sep = "\t", header = None)
reshaped_dat1 = dat1[2].values.reshape(11,11) # if your tsv file has 121 lines and 3 columns where the last column has counts. 
mat1 = csr_matrix(np.asarray(reshaped_dat1))
kr = krbalancing.kr_balancing(mat1.shape[0], mat1.shape[1],
                              mat1.count_nonzero(),
                              mat1.indptr.astype(np.int64, copy=False),
                              mat1.indices.astype(np.int64, copy=False),
                              mat1.data.astype(np.float64, copy = False))
```

Keep in mind that the algorithm works only on square matrices and returns an upper triangular matrix.

**Comment by serine** [comment]
Right, I see. The main confusion was in the fact that I couldn't (easily?) convert from dense (COO) matrix into CSR matrix. I did try using `coo_matrix()` but no luck. And the second confusion was with `kr_balancing()` constructor. it is just too cryptic for me to understand.

It would have been very nice if `kr_balancing()` would simply take a matrix, something similar to [ICED package](http://members.cbio.mines-paristech.fr/~nvaroquaux/iced/tutorial/index.html) 

```python
>>> from iced import normalization
>>> normed = normalization.ICE_normalization(counts)
```
where `counts` is a full, sparse matrix

---

So the way I got this to work is:

- made full, sparse matrix elsewhere and stored in a file and then 

```python
mat1 = np.loadtxt(fn_in, delimiter = " ",  usecols=range(n_cols))
mat2 = csr_matrix(np.asarray(mat1))
```
 and wrote it back up as full matrix. 

```python
mat2_norm = kr.get_normalised_matrix(True)
np.savetxt(fn_out, mat2_norm.todense(), delimiter = " ")
```

I'm a little unsure why I have to call `mat2_norm.todense()` .  I thought that dense representation was three columns representation, but what I see in the file is full, sparse matrix `n x n`

---

Overall I've managed to solve my problem, thanks a lot for helping  !
